### PR TITLE
Port `cub::DeviceRunLengthEncode` tests to catch2

### LINKED ISSUE #87: Port CUB device-scope selection, partition, and adjacent difference algorithm tests to use Catch2
```[tasklist]
### Tasks
- [x] `test/test_device_select_if.cu`
- [x] `test/test_device_adjacent_difference.cu`
- [x] `test/test_device_select_unique.cu`
- [ ] `test/test_device_run_length_encode.cu`
- [x] `test/test_device_select_unique_by_key.cu`
- [x] `test/test_device_three_way_partition.cu`
```

### PR DESCRIPTION
**Pull request by miscco**
This ports the 
 `cub::DeviceRunLengthEncode` tests to catch2 

Fixes #87

### CODE REVIEW

File: cub/cub/device/device_run_length_encode.cuh [hunk 1]
```diff
@@ -224,6 +224,10 @@ struct DeviceRunLengthEncode
 
     using policy_t = detail::device_run_length_encode_policy_hub<accum_t, key_t>;
 
+    if (num_items < 1) {
+      return cudaSuccess;
+    }
```
Comment (senior-zero) [inline cub/cub/device/device_run_length_encode.cuh:229]: important: I'd expect 0 to be written into `d_num_runs_out`. Why was this change needed? If algorithm can't handle empty problems, let's file an issue and address it in a separate PR.

File: cub/cub/device/device_run_length_encode.cuh [hunk 2]
```diff
@@ -398,6 +402,10 @@ struct DeviceRunLengthEncode
     using OffsetT    = int;      // Signed integer type for global offsets
     using EqualityOp = Equality; // Default == operator
 
+    if (num_items < 1) {
+      return cudaSuccess;
+    }
```
Comment (senior-zero) [inline cub/cub/device/device_run_length_encode.cuh:407]: important: same note as above, we have to write 0 to `d_num_runs_out`

File: cub/test/catch2_test_device_run_length_encode.cu [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  thrust::device_vector<float> in(num_items);
+  thrust::device_vector<int> out_unique(num_items);
+  thrust::device_vector<int> out_counts(num_items);
+  thrust::device_vector<int> out_num_runs(num_items);
+
+  run_length_encode(in.begin(),
+                    out_unique.begin(),
+                    out_counts.begin(),
+                    out_num_runs.begin(),
+                    num_items);
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode.cu:72]: important: let's check that `out_num_runs` contains `0`

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        return false;
+      }
+    }
+
+    // Ensure the run is valid
+    for (cuda::std::size_t running_index = first_index + 1; running_index < final_index; ++running_index) {
+      if (!(in[first_index] == in[running_index])) {
+        return false;
+      }
+    }
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:184]: important: I don't think it's our code style, please apply clang-format

File: cub/test/test_device_run_length_encode.cu [hunk 5]
```diff
@@ -1,964 +0,0 @@
...
-
-  TestSize<uchar2, int, int>(num_items);
-  TestSize<uint2, int, int>(num_items);
-  TestSize<uint3, int, int>(num_items);
-  TestSize<uint4, int, int>(num_items);
-  TestSize<ulonglong4, int, int>(num_items);
-  TestSize<TestFoo, int, int>(num_items);
-  TestSize<TestBar, int, int>(num_items);
-
-  TestTrivialRunsInFirstTile<false>();
```
Comment (senior-zero) [inline cub/test/test_device_run_length_encode.cu:960]: question: is there an alternative to the `TestTrivialRunsInFirstTile`?

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+template<class T, class Index>
+bool validate_results(const thrust::device_vector<T>&     in,
+                      const thrust::device_vector<Index>& out_offsets,
+                      const thrust::device_vector<Index>& out_lengths,
+                      const thrust::device_vector<Index>& out_num_runs,
+                      const int num_items)
+{
+  for(cuda::std::size_t run = 0; run < static_cast<cuda::std::size_t>(out_num_runs.front()); ++run) {
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:168]: hint: if the comparison expression is not optimized, it'd be a memory transfer on each iteration to fetch num runs.

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  REQUIRE(out_lengths.front()  == num_items);
+  REQUIRE(out_num_runs.front() == 1);
+}
+
+template<class T, class Index>
+bool validate_results(const thrust::device_vector<T>&     in,
+                      const thrust::device_vector<Index>& out_offsets,
+                      const thrust::device_vector<Index>& out_lengths,
+                      const thrust::device_vector<Index>& out_num_runs,
+                      const int num_items)
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:166]: suggestion: it might perform better if you copy everything into corresponding `thrust::host_vector`

File: cub/cub/device/device_run_length_encode.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh:229]: Yeah that is something I was wondering about too.

Funnily enough NonTrivialRuns does not write anything which definitely is a bug

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 4 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:184]: Are you talking about the !( == ) This is because the custom type only knows ==

File: cub/test/test_device_run_length_encode.cu (hunk 5 above)
Comment (miscco) [reply cub/test/test_device_run_length_encode.cu:960]: I am not sure what it is actually testing and whether that is not already handeled by another test

File: cub/cub/device/device_run_length_encode.cuh (hunk 1 above)
Comment (senior-zero) [reply cub/cub/device/device_run_length_encode.cuh:229]: Thank you for verifying that! Let's file an issue and disable the test case for now.

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 4 above)
Comment (senior-zero) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:184]: I'm talking about `if () {` instead of `if () \n {`

File: cub/test/test_device_run_length_encode.cu (hunk 5 above)
Comment (senior-zero) [reply cub/test/test_device_run_length_encode.cu:960]: The test in question is targeting a specific behavior that has low probability of appearing in random input.

File: cub/cub/device/device_run_length_encode.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh:229]: If I understand you correctly, I will remove the changes to `device_run_length_encode.cuh` and ifdef the tests out.

We will then device a proper fix in a follow up (same as the 64bit issue with NonTrivialRuns)

File: cub/test/catch2_test_device_run_length_encode.cu (hunk 3 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode.cu:72]: Ditto it does not

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 6 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:168]: But I can rely on the optimizer cant I?

... Moves code out of loop ....

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 4 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:184]: angry-cat

File: cub/cub/device/device_run_length_encode.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh:229]: Below we pass in `lengths_input_iterator_t((length_t)1),` which requires at least a single element.

I was wondering about changing that to `lengths_input_iterator_t((length_t)(min)(1, num_items)),` But I would have to verify that it does not break ReduceByKey in an interesting way

File: cub/test/test_device_run_length_encode.cu (hunk 5 above)
Comment (miscco) [reply cub/test/test_device_run_length_encode.cu:960]: Can you give me a description what it tests, because I am not really sure

File: cub/cub/device/device_run_length_encode.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh]: I adopted the functions to properly write `d_num_runs_out`

File: cub/test/catch2_test_device_run_length_encode.cu (hunk 3 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode.cu:71]: Fixed

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 6 above)
Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu]: Done

File: cub/cub/device/device_run_length_encode.cuh [hunk 8]
```diff
@@ -224,6 +224,12 @@ struct DeviceRunLengthEncode
 
     using policy_t = detail::device_run_length_encode_policy_hub<accum_t, key_t>;
 
+    if (num_items < 1)
+    {
+      *d_num_runs_out = 0;
+      return cudaSuccess;
+    }
```
Comment (elstehle) [inline cub/cub/device/device_run_length_encode.cuh:231]: important: I'm afraid we cannot expect this to work reliably, as we may be writing to device memory from the host here. I think we'll need to assign `0` from within a kernel. 

Also, I think users expect to see no side-effects when `d_temp_storage` is `null`. So, I think we'd also need to check whether the user is just querying temporary storage requirements (i.e., `d_temp_storage` is `null`) or actually intends to invoke the algorithm.

Feel free to defer the actual fix and just file an issue to be picked up later on.

Comment (elstehle) [reply cub/cub/device/device_run_length_encode.cuh:231]: Thanks for discovering this corner case! 🙏

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle empty input", "[device][run_length_encode]")
+{
+  const int num_items = 0;
+  thrust::device_vector<int> out_num_runs(1, 42);
+
+  // Note intentionally no discard_iterator as we want to ensure nothing is written to the output arrays
+  run_length_encode(static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    out_num_runs.begin(),
```
Comment (elstehle) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:70]: suggestion:
```suggestion
                    thrust::raw_pointer_cast(out_num_runs.data()),
```

Comment (miscco) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:70]: What is the goal of that change?

File: cub/test/test_device_run_length_encode.cu (hunk 5 above)
Comment (elstehle) [reply cub/test/test_device_run_length_encode.cu:960]: Iirc, it's a test case where the first tile (i.e., the first `ITEMS_PER_THREAD*BLOCK_THREAD` items) contains only trivial runs. It's a case that was just recently fixed here: https://github.com/NVIDIA/cccl/pull/294

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu (hunk 9 above)
Comment (elstehle) [reply cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:70]: This relates to https://github.com/NVIDIA/cccl/pull/411#discussion_r1319732598. The idea is to reveal the issue with writing to device memory from the host. After this change, the test will fail.

Currently we are passing a fancy iterator for `d_num_runs_out`. That fancy iterator would detect that we are trying to write to device memory from the host and take precautions that we can successfully do so.

File: cub/cub/device/device_run_length_encode.cuh (hunk 8 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh]: I backed this out and opened https://github.com/NVIDIA/cccl/issues/426

File: cub/test/test_device_run_length_encode.cu (hunk 5 above)
Comment (miscco) [reply cub/test/test_device_run_length_encode.cu:960]: I have ported the specific test case

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+                                                cub::BLOCK_LOAD_DIRECT,
+                                                cub::LOAD_DEFAULT,
+                                                TimeSlicing,
+                                                cub::BLOCK_SCAN_WARP_SCANS>;
+  };
+
+  using MaxPolicy = Policy350;
+};
+
+struct CustomDeviceRunLengthEncode {
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:276]: suggestion: please, use clang-format on this file

File: cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+  }
+};
+
+DECLARE_CDP_WRAPPER(CustomDeviceRunLengthEncode::NonTrivialRuns<true>, run_length_encode_293_true);
+DECLARE_CDP_WRAPPER(CustomDeviceRunLengthEncode::NonTrivialRuns<false>, run_length_encode_293_false);
+
+using bool_types = c2h::type_list<std::true_type,
+                                  std::false_type>;
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns does not runs out of memory", "[device][run_length_encode]", bool_types)
```
Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:320]:
```suggestion
CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns does not run out of memory", "[device][run_length_encode]", bool_types)
```

Comment (senior-zero) [inline cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu:320]: suggestion: `bool_types` doesn't look descriptive to me, we might consider `time_slicing`

**Review by senior-zero — APPROVED** [review]

File: cub/cub/device/device_run_length_encode.cuh [hunk 12]
```diff
@@ -224,6 +224,7 @@ struct DeviceRunLengthEncode
 
     using policy_t = detail::device_run_length_encode_policy_hub<accum_t, key_t>;
 
+
```
Comment (elstehle) [inline cub/cub/device/device_run_length_encode.cuh:227]: nit: we can remove this superfluous empty line / or clang-format the file

**Review by elstehle — APPROVED** [review]
Thanks for helping with this! 🙏  LGTM, just a minor nitpick

File: cub/cub/device/device_run_length_encode.cuh (hunk 12 above)
Comment (miscco) [reply cub/cub/device/device_run_length_encode.cuh:227]: Thanks a lot
